cdap-app-fabric/src/main/java/io/cdap/cdap/internal/state/AppStateHandler.java

```diff
@@ -0,0 +1,117 @@
+import com.google.inject.Singleton;
+import io.cdap.cdap.common.NamespaceNotFoundException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.namespace.NamespaceQueryAdmin;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import java.nio.charset.StandardCharsets;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * Internal {@link HttpHandler} for Application State Management
+ */
+@Singleton
+@Path(Constants.Gateway.API_VERSION_3)
+public class AppStateHandler extends AbstractHttpHandler {
+  private static final Gson GSON = new GsonBuilder().create();
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final AppStateStore appStateStore;
+
+  @Inject
+  AppStateHandler(NamespaceQueryAdmin namespaceQueryAdmin,
+                  AppStateStore appStateStore) {
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.appStateStore = appStateStore;
+  }
+
+  /**
+   * Get {@link AppState} for a given app-name.
+   */
+  @GET
+  @Path("/namespaces/{namespace-id}/apps/{app-name}/appids/{app-id}/states/{state-key}")
+  public void getState(HttpRequest request, HttpResponder responder,
+                       @PathParam("namespace-id") String namespaceId,
+                       @PathParam("app-name") String appName,
+                       @PathParam("app-id") long appId,
+                       @PathParam("state-key") String stateKey) throws Exception {
+    validateNamespace(namespaceId);
+    AppState appStateRequest = new AppState(namespaceId, appName, appId, stateKey);
+    if (appStateStore.getState(appStateRequest).isPresent()) {
+      responder.sendJson(HttpResponseStatus.OK,
+              new String(appStateStore.getState(appStateRequest).get(), StandardCharsets.UTF_8));
+    } else {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+    }
+  }
+
+  /**
+   * Save {@link AppState} for a given app-name.
+   */
+  @POST
+  @Path("/namespaces/{namespace-id}/apps/{app-name}/appids/{app-id}/states/{state-key}")
+  public void saveState(FullHttpRequest request, HttpResponder responder,
+                        @PathParam("namespace-id") String namespaceId,
+                        @PathParam("app-name") String appName,
+                        @PathParam("app-id") long appId,
+                        @PathParam("state-key") String stateKey) throws Exception {
+    validateNamespace(namespaceId);
+    AppState appStateRequest = new AppState(namespaceId, appName, appId, stateKey,
+            request.content().toString(StandardCharsets.UTF_8).getBytes(StandardCharsets.UTF_8));
+    appStateStore.saveState(appStateRequest);
```

Comment:
this endpoint should actually return a 404 if the application doesn't exist. It will need to check the applications table in order to do this.

Usually we have a http handler layer that translates requests into java objects and validates the input. We then have a Service layer that performs the actual business logic, with the handler translating exceptions and response objects into http responses.

In this particular case, I think it makes sense to add methods to the ApplicationLifecycleService for writing/deleting state and have this handler call the Service instead of interacting with the store directly. I think we will also have to add methods to the Store interface for writing/deleting state instead of having a separate AppStateStore, as we will want to be able to delete app state and delete the app in the same transaction.